### PR TITLE
feat: add feature to record multi-host system

### PR DESCRIPTION
**Pull request by ymski**
## Description

This PR adds feature to record multi-host system.

## Related links

- https://github.com/tier4/caret_analyze/pull/455
- https://github.com/tier4/ros2caret/pull/163
- https://github.com/tier4/caret_doc/pull/190

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: CARET_trace/test/test_trace_node.cpp [hunk 1]
```diff
@@ -135,24 +142,36 @@ TEST(CaretTraceTest, TestStartTransition)
...
     EXPECT_TRUE(node_->get_status() == TRACE_STATUS::WAIT);
+    set_lttng_session_running_return_value(true);
+    auto start_msg = get_start_msg();
+    node_->start_callback(std::move(start_msg));
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+  }
 
+  {  // subscribe end message without lttng session: WAIT -> WAIT
+    set_lttng_session_running_return_value(false);
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:171]: I think this line can be removed.

File: CARET_trace/test/test_trace_node.cpp [hunk 2]
```diff
@@ -135,24 +142,36 @@ TEST(CaretTraceTest, TestStartTransition)
...
 
     EXPECT_TRUE(node_->get_status() == TRACE_STATUS::WAIT);
+    set_lttng_session_running_return_value(true);
+    auto start_msg = get_start_msg();
+    node_->start_callback(std::move(start_msg));
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+  }
 
+  {  // subscribe end message without lttng session: WAIT -> WAIT
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:170]: I suggest testing both "with lttng session" and "without lttng session".

File: CARET_trace/test/test_trace_node.cpp [hunk 3]
```diff
@@ -175,8 +195,33 @@ TEST_F(CaretTraceNodeTest, TestPrepareTransition)
...
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+
+    auto start_msg = get_start_msg();
+    node_->start_callback(std::move(start_msg));
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+  }
+
+  {  // subscribe end message with lttng session: PREPARE -> WAIT
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:210]: Minor fix of comment.

```suggestion
  {  // subscribe end message with lttng session: PREPARE -> PREPARE
```

File: CARET_trace/test/test_trace_node.cpp [hunk 4]
```diff
@@ -175,8 +195,33 @@ TEST_F(CaretTraceNodeTest, TestPrepareTransition)
...
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+
+    auto end_msg = get_end_msg();
+    node_->end_callback(std::move(end_msg));
+
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+  }
+
+  {  // subscribe end message without lttng session: PREPARE -> PREPARE
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:222]: Minor fix of comment.

```suggestion
  {  // subscribe end message without lttng session: PREPARE -> WAIT
```

File: CARET_trace/test/test_trace_node.cpp [hunk 5]
```diff
@@ -207,27 +252,49 @@ TEST_F(CaretTraceNodeTest, TestMeasureTransition)
 {
   init_context();
 
-  {  // subscribe start message: MEASURE -> PREPARE
+  {  // subscribe start message with lttng session: MEASURE -> PREPARE
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:255]: Minor fix of comment. (Let's change MEASURE to RECORD on this occasion)
```suggestion
  {  // subscribe start message with lttng session: RECORD -> RECORD
```

File: CARET_trace/test/test_trace_node.cpp [hunk 6]
```diff
@@ -207,27 +252,49 @@ TEST_F(CaretTraceNodeTest, TestMeasureTransition)
...
+    set_lttng_session_running_return_value(true);
     auto start_msg = get_start_msg();
     node_->start_callback(std::move(start_msg));
 
-    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::PREPARE);
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::RECORD);
   }
 
-  {  // subscribe end message: MEASURE -> WAIT
+  {  // subscribe start message without lttng session: MEASURE -> PREPARE
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:266]: Minor fix of comment.
```suggestion
  {  // subscribe start message without lttng session: RECORD -> RECORD
```

File: CARET_trace/test/test_trace_node.cpp [hunk 7]
```diff
@@ -207,27 +252,49 @@ TEST_F(CaretTraceNodeTest, TestMeasureTransition)
...
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::RECORD);
+
+    set_lttng_session_running_return_value(false);
+    auto start_msg = get_start_msg();
+    node_->start_callback(std::move(start_msg));
 
     EXPECT_TRUE(node_->get_status() == TRACE_STATUS::RECORD);
+  }
+
+  {  // subscribe end message without lttng session: MEASURE -> WAIT
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:277]: Minor fix of comment.
```suggestion
  {  // subscribe end message without lttng session: RECORD -> WAIT
```

File: CARET_trace/test/test_trace_node.cpp [hunk 8]
```diff
@@ -207,27 +252,49 @@ TEST_F(CaretTraceNodeTest, TestMeasureTransition)
...
+    EXPECT_TRUE(node_->get_status() == TRACE_STATUS::RECORD);
 
+    set_lttng_session_running_return_value(false);
     auto end_msg = get_end_msg();
     node_->end_callback(std::move(end_msg));
 
     EXPECT_TRUE(node_->get_status() == TRACE_STATUS::WAIT);
   }
+
+  {  // subscribe end message with lttng session: MEASURE -> WAIT
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:288]: Minor fix of comment.

```suggestion
  {  // subscribe end message with lttng session: RECORD -> RECORD
```

File: CARET_trace/test/test_trace_node.cpp [hunk 9]
```diff
@@ -244,6 +311,7 @@ TEST_F(CaretTraceNodeTest, TestPrepare)
   init_context();
 
   set_status(TRACE_STATUS::PREPARE);
+  set_lttng_session_running_return_value(true);
```
Comment (isp-uetsuki) [inline CARET_trace/test/test_trace_node.cpp:314]: Is this change necessary?

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

File: CARET_trace/CMakeLists.txt [hunk 10]
```diff
@@ -62,7 +62,7 @@ ament_target_dependencies(caret
 ament_export_targets(export_caret_trace HAS_LIBRARY_TARGET)
 ament_export_dependencies(rclcpp LTTngUST)
 
-target_link_libraries(caret CycloneDDS::ddsc fastrtps)
+target_link_libraries(caret CycloneDDS::ddsc fastrtps uuid)
```
[the remark this answers is not in the document]
Comment (ymski) [reply CARET_trace/CMakeLists.txt:65]: Thanks for reviewing.
This issue has been fixed in [945dd48](https://github.com/tier4/caret_trace/pull/262/commits/945dd48f7f9cf940c0ba541ad1a80973fbc484a4)

File: CARET_trace/src/trace_node.cpp [hunk 11]
```diff
@@ -87,11 +89,27 @@ TraceNode::~TraceNode()
...
+  return base_name;
+}
+
+std::string TraceNode::get_formatted_uuid()
+{
+  uuid_t uuid_value;
+  std::unique_ptr<char[]> uuid_string_ptr = std::make_unique<char[]>(UUID_STR_LEN);
+  uuid_generate(uuid_value);
+  uuid_unparse_lower(uuid_value, uuid_string_ptr.get());
+  auto base_name = std::string(uuid_string_ptr.get());
```
[the remark this answers is not in the document]
Comment (ymski) [reply CARET_trace/src/trace_node.cpp]: fixed in [14a25d6](https://github.com/tier4/caret_trace/pull/262/commits/14a25d6401dbd923e545f831f53846c9848feced).

File: CARET_trace/test/test_trace_node.cpp (hunk 1 above)
Comment (ymski) [reply CARET_trace/test/test_trace_node.cpp:172]: Fixed in [f53bfac](https://github.com/tier4/caret_trace/pull/262/commits/f53bfac8e8e37f47b05e2e43822018e8433cd85e).
This commit reflects the all comments below.

**Comment by isp-uetsuki** [comment]
LGTM

**Review by isp-uetsuki — APPROVED** [review]

**Review by xygyo77 — APPROVED** [review]
LGTM
